deploy-service/common/src/main/java/com/pinterest/deployservice/handler/GoalAnalyst.java

```diff
@@ -368,22 +377,38 @@ AgentBean genUpdateBeanByReport(PingReportBean report, AgentBean agent) {
         if (agent == null) {
             // if agent is missing in agent table, treat it as not first_deploy.
             updateBean.setFirst_deploy(false);
-            updateBean.setStart_date(System.currentTimeMillis());
+            updateBean.setStart_date(currentTime);
         } else {
             updateBean.setFirst_deploy(agent.getFirst_deploy());
             updateBean.setStart_date(agent.getStart_date());
         }
 
-        if (report.getDeployStage() == DeployStage.SERVING_BUILD) {
+        if (report.getDeployStage() == DeployStage.SERVING_BUILD && updateBean.getFirst_deploy()) {
             // turn off first deploy flag
             updateBean.setFirst_deploy(false);
-            updateBean.setFirst_deploy_time(System.currentTimeMillis());
+            updateBean.setFirst_deploy_time(currentTime);
+            emitMetrics(updateBean);
         }
 
         // TODO record error message as well if errorCode != 0
         return updateBean;
     }
 
+    private void emitMetrics(AgentBean updateBean) {
+        try {
+            EnvironBean env = envs.get(updateBean.getEnv_id());
+            Metrics.timer(String.format(DEPLOY_LATENCY_TIMER_NAME, env.getEnv_name(), env.getStage_name()))
+                    .record(Duration.ofMillis(updateBean.getFirst_deploy_time() - updateBean.getStart_date()));
+            Metrics.counter(FIRST_DEPLOY_COUNTER_NAME, "success",
+                    String.valueOf(updateBean.getStatus().equals(AgentStatus.SUCCEEDED)),
+                    "env", env.getEnv_name(),
+                    "stage", env.getStage_name())
```

Comment:
Any reason to set the env/stage in the metric name instead of the tags? Looks like both approaches are used here

Reply:
That's a good question. The reason is that the metrics is moved from Rodimus, and it has env and stage as tags there. Since the counter name will be changed anyways, I think it's better to have the env and stage in the name.